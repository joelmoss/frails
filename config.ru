# frozen_string_literal: true

require 'rubygems'
require 'bundler'

Bundler.require :default, :development

Combustion.path = 'test/dummy'
Combustion.initialize! :action_controller, :action_view do
  config.hosts << 'localhost'
end

run Combustion::Application
