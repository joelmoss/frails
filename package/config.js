const spawnSync = require('child_process').spawnSync
const path = require('path')

const rootPath = process.cwd()
const target = { rootPath, appPath: path.join(rootPath, 'app') }
const handler = {
  get: function (target, prop, receiver) {
    // Return the prop if it already exists.
    if (Object.keys(target).includes(prop)) return target[prop]

    // Return the prop if found in the Rails config cache.
    const config = getConfigFromRails()
    if (Object.keys(config).includes(prop)) return (target[prop] = config[prop])

    if (prop === 'absolutePublicPath') {
      return (target[prop] = path.join(rootPath, 'public', config.publicOutputPath))
    }

    return Reflect.get(...arguments)
  }
}
const config = new Proxy(target, handler)

module.exports = config

function getConfigFromRails() {
  if (configFromRails) return configFromRails

  const ioDelimiter = '__FRAILS_RAILS_RUNNER_DELIMETER__'

  const result = spawnSync('./bin/rails', [
    'runner',
    `puts "${ioDelimiter}#{Frails.config_as_json}${ioDelimiter}"`
  ])

  if (result.status === 0) {
    // Output is delimited to filter out unwanted warnings or other output
    // that we don't want in our files.
    const sourceRegex = new RegExp(ioDelimiter + '([\\s\\S]+)' + ioDelimiter)
    const matches = result.stdout.toString().match(sourceRegex)
    if (matches) {
      configFromRails = JSON.parse(matches[1])
    } else {
      throw 'Rails runner failed having been unable to parse the output.'
    }
  } else if (result.signal !== null) {
    throw `Rails runner was terminated with signal: ${result.signal}`
  } else if (result.error) {
    throw `Rails runner failed with '${result.error.toString()}'.`
  } else {
    throw `Rails runner failed with '${result.stderr.toString()}'.`
  }

  return configFromRails
}

// Cache the rails config.
let configFromRails = undefined
