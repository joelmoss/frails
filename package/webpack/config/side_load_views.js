const { isPlainObject, has } = require('lodash')
const { merge } = require('webpack-merge')
const glob = require('fast-glob')
const path = require('path')

const { appPath } = require('../../config')

module.exports = baseConfig => {
  if (typeof baseConfig.sideLoadViews === 'undefined') {
    // default
    baseConfig.sideLoadViews = true
  }

  if (baseConfig.sideLoadViews) {
    baseConfig.entry = merge(baseConfig.entry, entry())
  }

  delete baseConfig.sideLoadViews
}

// Entry points for side-loaded CSS and JS, which are any .css and .js files that exist alongside
// view and partial templates. For example, a view template at /app/views/pages/show.html.erb could
// have a css file alongside it at /app/views/pages/show.css.
//
// NOTE: Adding new files will require a rebuild!
function entry() {
  const result = {}

  glob.sync(path.join(appPath, 'views/**/*.{css,js}')).forEach(paf => {
    const namespace = path.relative(appPath, path.dirname(paf))
    const name = path.join(namespace, path.basename(paf, path.extname(paf)))

    has(result, name) ? result[name].push(paf) : (result[name] = [paf])
  })

  return result
}
