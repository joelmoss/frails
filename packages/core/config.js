const { spawnSync } = require('child_process')
const path = require('path')

const rootPath = process.cwd()
const target = { rootPath, appPath: path.join(rootPath, 'app') }
const handler = {
  get: function (target, prop) {
    // Return the prop if it already exists.
    if (Object.keys(target).includes(prop)) return target[prop]

    // Return the prop if found in the Rails config cache.
    const config = getConfigFromRails()
    if (Object.keys(config).includes(prop)) return (target[prop] = config[prop])

    if (prop === 'absolutePublicPath') {
      return (target.absolutePublicPath = path.join(rootPath, 'public', config.publicOutputPath))
    }

    if (prop === 'cssLocalIdentName') {
      // The local ident name required for loading component styles.
      target.cssLocalIdentName =
        config.railsEnv == 'development'
          ? '[path][name]__[local]___[md5:hash:hex:6]'
          : '[local]-[md5:hash:hex:6]'
    }

    return Reflect.get(...arguments)
  }
}

module.exports = new Proxy(target, handler)

function getConfigFromRails() {
  if (configFromRails) return configFromRails

  const ioDelimiter = '__FRAILS_RAILS_RUNNER_DELIMETER__'

  const result = spawnSync('./bin/rails', [
    'runner',
    `puts "${ioDelimiter}#{Frails.config_as_json}${ioDelimiter}"`
  ])

  if (result.status === 0) {
    // Output is delimited to filter out unwanted warnings or other output
    // that we don't want in our files.
    const sourceRegex = new RegExp(ioDelimiter + '([\\s\\S]+)' + ioDelimiter)
    const matches = result.stdout.toString().match(sourceRegex)
    if (matches) {
      configFromRails = JSON.parse(matches[1])
    } else {
      throw 'Rails runner failed having been unable to parse the output.'
    }
  } else if (result.signal !== null) {
    throw `Rails runner was terminated with signal: ${result.signal}`
  } else if (result.error) {
    throw `Rails runner failed with '${result.error.toString()}'.`
  } else {
    throw `Rails runner failed with '${result.stderr.toString()}'.`
  }

  return configFromRails
}

// Cache the rails config.
let configFromRails = undefined
