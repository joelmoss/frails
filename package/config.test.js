import config from "./config";

describe("config", () => {
  test("config", () => {
    expect(config).toMatchSnapshot();
  });
});
