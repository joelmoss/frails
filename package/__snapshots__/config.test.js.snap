// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config config 1`] = `
Object {
  "absolutePublicPath": "/Users/joelmoss/dev/frails/test/dummy/public/frails",
  "appPath": "/Users/joelmoss/dev/frails/test/dummy/app",
  "publicOutputPath": "frails",
  "rootPath": "/Users/joelmoss/dev/frails/test/dummy",
}
`;
